File: hanimo_rag/core/chunker.py
"""Recursive text chunker."""
from __future__ import annotations

from dataclasses import dataclass


@dataclass
class TextChunk:
    """A chunk of text with position info."""

    text: str
    start: int
    end: int
    index: int


# Separator hierarchy: split by largest units first
_DEFAULT_SEPARATORS = ["\n\n", "\n", ". ", ", ", " ", ""]


def chunk_text(
    text: str,
    chunk_size: int = 512,
    chunk_overlap: int = 50,
    separators: list[str] | None = None,
) -> list[TextChunk]:
    """Split text into overlapping chunks using recursive separator hierarchy."""
    if not text.strip():
        return []

    if separators is None:
        separators = _DEFAULT_SEPARATORS

    raw_splits = _recursive_split(text, chunk_size, separators)
    return _merge_with_overlap(raw_splits, chunk_size, chunk_overlap)


def _recursive_split(
    text: str,
    chunk_size: int,
    separators: list[str],
) -> list[str]:
    """Recursively split text using separator hierarchy."""
    if len(text) <= chunk_size:
        return [text] if text.strip() else []

    if not separators:
        # Last resort: hard split by characters
        pieces = []
        for i in range(0, len(text), chunk_size):
            piece = text[i : i + chunk_size]
            if piece.strip():
                pieces.append(piece)
        return pieces

    separator = separators[0]
    remaining_separators = separators[1:]

    if not separator:
        # Empty separator = character-level split
        pieces = []
        for i in range(0, len(text), chunk_size):
            piece = text[i : i + chunk_size]
            if piece.strip():
                pieces.append(piece)
        return pieces

    parts = text.split(separator)
    results: list[str] = []
    current = ""

    for part in parts:
        candidate = f"{current}{separator}{part}" if current else part

        if len(candidate) <= chunk_size:
            current = candidate
        else:
            if current.strip():
                results.append(current)
            # If this single part is too big, recursively split it
            if len(part) > chunk_size:
                sub_parts = _recursive_split(part, chunk_size, remaining_separators)
                results.extend(sub_parts)
                current = ""
            else:
                current = part

    if current.strip():
        results.append(current)

    return results


def _merge_with_overlap(
    splits: list[str],
    chunk_size: int,
    overlap: int,
) -> list[TextChunk]:
    """Merge splits into chunks with overlap."""
    if not splits:
        return []

    chunks: list[TextChunk] = []
    position = 0

    for i, text in enumerate(splits):
        # Add overlap from previous chunk
        if i > 0 and overlap > 0:
            prev_text = splits[i - 1]
            overlap_text = prev_text[-overlap:] if len(prev_text) > overlap else prev_text
            text_with_overlap = overlap_text + text
            start = max(0, position - len(overlap_text))
        else:
            text_with_overlap = text
            start = position

        # Trim if too long
        if len(text_with_overlap) > chunk_size:
            text_with_overlap = text_with_overlap[:chunk_size]

        chunks.append(
            TextChunk(
                text=text_with_overlap.strip(),
                start=start,
                end=start + len(text_with_overlap),
                index=i,
            )
        )

        position += len(text)

    # Filter out empty chunks
    return [c for c in chunks if c.text]

File: js/src/core/chunker.ts
/**
 * hanimo-rag v2 — Recursive Text Chunker
 *
 * Splits text into overlapping chunks using a hierarchy of separators,
 * similar to LangChain's RecursiveCharacterTextSplitter.
 */

const DEFAULT_SEPARATORS = [
  '\n\n\n',   // triple newline (section breaks)
  '\n\n',     // paragraph breaks
  '\n',       // line breaks
  '. ',       // sentence breaks
  '? ',       // question breaks
  '! ',       // exclamation breaks
  '; ',       // semicolon breaks
  ', ',       // comma breaks
  ' ',        // word breaks
]

export interface ChunkerOptions {
  chunkSize: number
  chunkOverlap: number
  separators?: string[]
}

export function chunkText(
  text: string,
  options: ChunkerOptions = { chunkSize: 1000, chunkOverlap: 200 },
): string[] {
  const { chunkSize, chunkOverlap, separators = DEFAULT_SEPARATORS } = options

  if (!text || text.trim().length === 0) {
    return []
  }

  if (text.length <= chunkSize) {
    return [text.trim()]
  }

  return recursiveSplit(text, chunkSize, chunkOverlap, separators)
}

function recursiveSplit(
  text: string,
  chunkSize: number,
  chunkOverlap: number,
  separators: string[],
): string[] {
  const chunks: string[] = []

  // Find the best separator for this text
  let bestSep = ''
  for (const sep of separators) {
    if (text.includes(sep)) {
      bestSep = sep
      break
    }
  }

  // If no separator found, split by character position
  if (!bestSep) {
    return splitBySize(text, chunkSize, chunkOverlap)
  }

  const parts = text.split(bestSep)
  let currentChunk = ''

  for (const part of parts) {
    const candidate = currentChunk
      ? currentChunk + bestSep + part
      : part

    if (candidate.length <= chunkSize) {
      currentChunk = candidate
    } else {
      // Current chunk is full
      if (currentChunk) {
        const trimmed = currentChunk.trim()
        if (trimmed.length > 0) {
          chunks.push(trimmed)
        }
      }

      // If this single part exceeds chunk size, recurse with next separator level
      if (part.length > chunkSize) {
        const remainingSeps = separators.slice(separators.indexOf(bestSep) + 1)
        if (remainingSeps.length > 0) {
          const subChunks = recursiveSplit(part, chunkSize, chunkOverlap, remainingSeps)
          chunks.push(...subChunks)
        } else {
          chunks.push(...splitBySize(part, chunkSize, chunkOverlap))
        }
        currentChunk = ''
      } else {
        currentChunk = part
      }
    }
  }

  // Don't forget the last chunk
  if (currentChunk.trim().length > 0) {
    chunks.push(currentChunk.trim())
  }

  // Apply overlap: merge small trailing/leading content between chunks
  return applyOverlap(chunks, chunkOverlap)
}

function splitBySize(text: string, chunkSize: number, chunkOverlap: number): string[] {
  const chunks: string[] = []
  let start = 0

  while (start < text.length) {
    const end = Math.min(start + chunkSize, text.length)
    const chunk = text.slice(start, end).trim()
    if (chunk.length > 0) {
      chunks.push(chunk)
    }
    start = end - chunkOverlap
    if (start >= text.length || end === text.length) break
  }

  return chunks
}

function applyOverlap(chunks: string[], overlap: number): string[] {
  if (chunks.length <= 1 || overlap <= 0) {
    return chunks
  }

  const result: string[] = []

  for (let i = 0; i < chunks.length; i++) {
    let chunk = chunks[i] ?? ''

    if (i > 0) {
      // Prepend overlap from previous chunk
      const prev = chunks[i - 1] ?? ''
      if (prev.length > overlap) {
        const overlapText = prev.slice(-overlap)
        // Only prepend if it doesn't make the chunk too large
        if (chunk.length + overlapText.length <= chunk.length * 2) {
          chunk = overlapText + ' ' + chunk
        }
      }
    }

    result.push(chunk.trim())
  }

  return result
}
